docs/rfc/rfc-7.adoc

```diff
@@ -0,0 +1,288 @@
+client works on a 2-core machine and there are no more than 5 signing group
+members on a single machine, the bottleneck phase should take no more than
+1 minute.
+
+The announcement phase takes always 7 blocks, so ~1.5 minutes assuming 12s block
+time. With 30 blocks timeout for a single attempt, it leaves more than 3 minutes
+for the rest of the signing protocol which should be more than enough.
+This should also be enough for local development when all signing group members
+reside on the same computer. Based on the local benchmarks, the entire signing
+never takes more than 5 minutes locally, and usually, it takes less than 4
+minutes.
+
+Assuming it takes no more than 4 minutes in the happy path to sign a single
+input, it should take no more than an hour to sign 15 inputs. Given that the
+entire sweep operation - including 6 Bitcoin confirmations - should complete
+within 4 hours, it gives just enough time for a single Bitcoin miner fee bump.
+
+==== Retries
+
+
+The signing protocol is non-attributable and we need to retry in case of
+corrupted data:
+
+- With 1 malicious member in a signing group, we need 2 attempts of the protocol
+in the worst case (`P = (99 choose 51) / (100 choose 51) = 0.49`).
+In the worst case, it takes approximately 2 hours to sign 15 inputs.
+
+- With 2 malicious members in a signing group, we need 5 attempts of the protocol
+in the worst case (`P = (98 choose 51) / (100 choose 51) = 0.2375757575`).
+In the worst case, it takes approximately 7h 30min to sign 15 inputs.
+
+- With 3 malicious members in a signing group, we need 10 attempts of the protocol
+in the worst case (`P = (97 choose 51) / (100 choose 51) = 0.1139393939`).
+In the worst case, it takes approximately 15 hours to sign 15 inputs.
+
+- With 4 malicious members in a signing group, we need 20 attempts of the
+protocol in the worst case (`P = (96 choose 51) / (100 choose 51) = 0.0540331146`).
+In the worst case, it takes approximately 30 hours to sign 15 inputs.
+
+- With 5 malicious members in a signing group, we need 40 attempts of the protocol
+in the worst case (`P = (95 choose 51) / (100 choose 51) = 0.0253280224`).
+In the worst case, it takes approximately 60 hours to sign 15 inputs.
+
+This attack slows down the sweeping schedule but given that the active wallet is
+not performing redemptions or moving funds, it does not affect the funds already
+bridged.
+
+In the first release, the client should assume no more than 3% of the network is
+corrupted and retry the signing of a single input no more than 10 times. If the
+signing of an input failed 10 times, the entire sweep is stopped. This puts the
+group in danger of a fraud challenge from one of the group members but the group
+will have a chance to sweep the deposit during the next sweep window. To defeat
+the fraud challenge, it is enough that the deposit has been swept, and it does
+not matter in which particular transaction.
```

Comment:
Can you elaborate on the fraud challenge here?

Reply:
Frauds work in a challenge and defeat-the-challenge mode. The challenger [provides a signature and `preimageSha256`](https://github.com/keep-network/tbtc-v2/blob/50267dbf9df4427f475544e459e7ce5a8aa0284a/solidity/contracts/bridge/Fraud.sol#L134-L139). If the wallet signed a transaction, submitted it to the mempool, and after some time decided to bump up the miner fee providing signature over another transaction, the first signature can be used for a challenge. Since that transaction does not exist on the chain, we can't look it up there. Actually, I am not even sure if such a lookup would work for mined transaction. In other words, the wallet must store all successful signatures locally, regardless if they were for transactions that got mined or for transactions that got replaced.

Reply:
Roger - I remember us talking about this a few months ago. I don't think it's worth explaining this inline, but I do think it would be useful as a footnote or something. What do you think?

Reply:
Sure! Please see bf716a4ff0b4dab3f19974df41498d69ef74e6fd.